### PR TITLE
Refactor workflow

### PR DESCRIPTION
**Pull request by mahesh-panchal**
- Adds functions for common map operations.
- Use two channels to structure inputs and outputs of each workflow stage.
- Move certain parts around to outside stages ( need to control with other parameter flags ).
- Attempt to simplify reading of primary workflow.
- Adds output channel for common join

### CODE REVIEW

**Comment by mahesh-panchal** [comment]
What's your opinion on this refactor? 
Is it more readable?
Can you tell what's going on in the `main.nf` much better?

**Comment by MartinPippel** [comment]
I really like all changes! 
- the additional assembly auxiliary functions are pretty helpful (now that I got what they are supposed to do). 
- If I get this if-else logic right, then this keeps the automatic flow intact, e.g. if a step is removed from the pipeline?

Meaning, if I omit the decontamination step e.g.  via`"steps": "inspect,preprocess,assemble,purge,polish"` then the following code snippet just assigns the `ch_raw_assemblies` channel to the `ch_cleaned_assemblies` channel, right?. Thats cool!!! 

```bash 
    // Contamination screen
    ch_to_screen = setAssemblyStage (
        ch_raw_assemblies,
        'decontaminated' // Set assembly stage now for filenaming
    ).dump(tag: 'Assemblies: to screen')
    if ( 'screen' in workflow_steps ) {
        DECONTAMINATE( ch_to_screen )
        ch_cleaned_assemblies = DECONTAMINATE.out.assemblies
    } else {
        ch_cleaned_assemblies = ch_to_screen
    }
    ch_cleaned_assemblies = ch_cleaned_assemblies.mix (
        preassembledInput( PREPARE_INPUT.out.assemblies, 'decontaminated' )
    ).dump(tag: 'Assemblies: Cleaned')
    ```
Have you tested it on a real example already? If not I can give it a try on a 50Mb genome.

**Comment by mahesh-panchal** [comment]
> I really like all changes!
> 
>     * the additional assembly auxiliary functions are pretty helpful (now that I got what they are supposed to do).
> 
👍🏽 

>     * If I get this if-else logic right, then this keeps the automatic flow intact, e.g. if a step is removed from the pipeline?

Exactly

> Meaning, if I omit the decontamination step e.g. via`"steps": "inspect,preprocess,assemble,purge,polish"` then the following code snippet just assigns the `ch_raw_assemblies` channel to the `ch_cleaned_assemblies` channel, right?. Thats cool!!!
> 
> ```shell
>     // Contamination screen
>     ch_to_screen = setAssemblyStage (
>         ch_raw_assemblies,
>         'decontaminated' // Set assembly stage now for filenaming
>     ).dump(tag: 'Assemblies: to screen')
>     if ( 'screen' in workflow_steps ) {
>         DECONTAMINATE( ch_to_screen )
>         ch_cleaned_assemblies = DECONTAMINATE.out.assemblies
>     } else {
>         ch_cleaned_assemblies = ch_to_screen
>     }
>     ch_cleaned_assemblies = ch_cleaned_assemblies.mix (
>         preassembledInput( PREPARE_INPUT.out.assemblies, 'decontaminated' )
>     ).dump(tag: 'Assemblies: Cleaned')
> ```

Yes, and the stage information is updated automatically too

> 
> Have you tested it on a real example already? If not I can give it a try on a 50Mb genome.

Not yet. Please do try, but you may encounter syntax errors.

File: modules/local/functions.nf [hunk 1]
```diff
@@ -61,3 +57,23 @@ def joinByMetaKeys( Map args = [:], lhs, rhs ) {
             assert args.meta in ['lhs','rhs','merge','subset']
     }
 }
+
+def assembliesFromStage( assemblies, String stage ) {
+    assemblies.filter { meta, assembly -> meta.assembly?.stage == stage }
+}
+
+def setAssemblyStage( assemblies, String stage ) {
+    assemblies.map{ meta, assemblies -> [ meta.deepMerge([ assembly: [ stage: stage, build: "${meta.assembly.assembler}-${stage}-${meta.assembly.id}" ] ]), assemblies ] }
```
Comment (MartinPippel) [inline modules/local/functions.nf:66]: I get following error: 

N E X T F L O W  ~  version 23.10.1
WARN: It appears you have never run this project before -- Option `-resume` is ignored
Launching `/home/pippel/tests/Earth-Biogenome-Project-pilot/main.nf` [golden_yalow] DSL2 - revision: b3d3024b05
ERROR ~ Module compilation error
- file : /home/pippel/tests/Earth-Biogenome-Project-pilot/modules/local/functions.nf
- cause: The current parameter list already contains a parameter of the name assemblies
 @ line 66, column 19.
       assemblies.map{ meta, assemblies -> [ meta.deepMerge([ assembly: [ stage: stage, build: "${meta.assembly.assembler}-${stage}-${meta.assembly.id}" ] ]), assemblies ] }
      

Its probably just a typo:

```suggestion
    assemblies.map{ meta, assembly -> [ meta.deepMerge([ assembly: [ stage: stage, build: "${meta.assembly.assembler}-${stage}-${meta.assembly.id}" ] ]), assemblies ] }
```

File: main.nf [hunk 2]
```diff
@@ -3,31 +3,26 @@
...
-include { ASSEMBLE_HIFI } from "$projectDir/subworkflows/local/assemble_hifi/main"
+include { ASSEMBLE                                   } from "$projectDir/subworkflows/local/assemble/main"
+include { ASSEMBLE_ORGANELLES                        } from "$projectDir/subworkflows/local/assemble_organelles/main"
+include { COMPARE_ASSEMBLIES                         } from "$projectDir/subworkflows/local/compare_assemblies/main"
+include { EVALUATE_ASSEMBLY as EVALUATE_RAW_ASSEMBLY } from "$projectDir/subworkflows/local/evaluate_assembly/main"
 
-include { FCSGX_FETCHDB } from "$projectDir/modules/local/fcsgx/fetchdb"
-include { FCSGX_RUNGX   } from "$projectDir/modules/local/fcsgx/rungx"
-include { FCSGX_CLEAN   } from "$projectDir/modules/local/fcsgx/clean"
+include { DECONTAMINATE } from "$projectDir/modules/local/decontaminate/main"
```
Comment (MartinPippel) [inline main.nf:22]:
```suggestion
include { DECONTAMINATE } from "$projectDir/subworkflows/local/decontaminate/main"
```

File: subworkflows/local/inspect_data/main.nf [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    hifi_histogram // [ meta, hifi_hist, ktab ]
+    hic_histogram  // [ meta, hic_hist, ktab ]
+
+    main:
+    ch_versions = Channel.empty()
+    // QC Steps
+    GENOME_PROPERTIES ( hifi_histogram )
+    COMPARE_LIBRARIES ( hifi_histogram.join( hic_histogram ) )
+    SCREEN_READS (
+        PREPARE_INPUT.out.hifi,
```
Comment (MartinPippel) [inline subworkflows/local/inspect_data/main.nf:18]:
```suggestion
        INSPECT_DATA.out.hifi,
```

Comment (MartinPippel) [reply subworkflows/local/inspect_data/main.nf:18]: nope, that suggestion is not working. Need to check further. 

I get this error:

`Launching `/home/pippel/tests/Earth-Biogenome-Project-pilot/main.nf` [silly_magritte] DSL2 - revision: 530b58d756

    Running NBIS Earth Biogenome Project Assembly workflow.

ERROR ~ No such variable: PREPARE_INPUT

 -- Check script '/home/pippel/tests/Earth-Biogenome-Project-pilot/subworkflows/local/inspect_data/main.nf' at line: 17 or see '.nextflow.log' file for more details
`

Comment (mahesh-panchal) [reply subworkflows/local/inspect_data/main.nf]: All channels needed within a workflow should be defined by either `take:` or created in the `main:`

**Comment by MartinPippel** [comment]
I got following error: 

```
ERROR ~ No such variable: meta

 -- Check script '/home/pippel/tests/Earth-Biogenome-Project-pilot/subworkflows/local/assemble_hifi/main.nf' at line: 42 or see '.nextflow.log' file for more details
```

But the file `subworkflows/local/assemble_hifi/main.nf` was not part of this PR. Would you mind to have a look? 

I potentially solved(?) this via changing this from 
```
        fasta_ch = GFATOOLS_GFA2FA.out.fasta.multiMap{
            fasta: [ meta, fasta ]
            genome_size: meta.sample.genome_size
```
into:

```
        fasta_ch = GFATOOLS_GFA2FA.out.fasta.multiMap{ meta, fasta ->
            fasta: [ meta, fasta ]
            genome_size: meta.sample.genome_size
```

**Comment by MartinPippel** [comment]
After fixing the previous error the pipeline finished successfully (FCS was excluded for time reasons)

**Comment by mahesh-panchal** [comment]
> But the file `subworkflows/local/assemble_hifi/main.nf` was not part of this PR. 

You should also be able to push commits to this branch.

**Comment by mahesh-panchal** [comment]
There's still more to update but we can clean up later.

**Review by MartinPippel — APPROVED** [review]
👍

**Pull request merged**